Add extra metric relabelings to scrape classes

sathieu opened the pull request:
## Description

See https://github.com/prometheus-operator/prometheus-operator/issues/5947#issuecomment-1845611166

Inspired by #6379.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
<!-- How you tested it? How do you know it works? -->
Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```release-note
Add extra metric relabelings to scrape classes
```

simonpasquier reviewed (commented):
The main issue for me is that the scrape class metrics relabeling happens after the scrape object's metrics relabeling. Other scrape class fields behave differently and can be overridden by scrape objects. I suppose that you have a specific requirement in mind but the inconsistency brings a UX problem.

sathieu commented:
@simonpasquier Yes. This is described in the doc, but if this is not enough we can use two configurations: `metricRelabelingsBefore` and `metricRelabelingsAfter` (I'm not good at naming)., or `metricRelabelings` and `metricRelabelingsOverride`, ....

simonpasquier commented:
Can you share more details about why you need scrape class relabeling to happen after service/pod monitor relabelings rather than before?

sathieu commented:
Yes.

As described [in the issue](https://github.com/prometheus-operator/prometheus-operator/issues/5947#issuecomment-1845611166), we have multi-tenant clusters, and each tenant is defined by a list of namespaces or a list of namespace regexps. We want the metrics relabeling to happen after the namespace label enforcement, to avoid a tenant injecting metrics into another tenant.

If scrape class metric relabelings happen before scrape resource metric relabelings, a malicious actor can easily change the `tenant` label.

simonpasquier commented:
> As described https://github.com/prometheus-operator/prometheus-operator/issues/5947#issuecomment-1845611166, we have multi-tenant clusters, and each tenant is defined by a list of namespaces or a list of namespace regexps. We want the metrics relabeling to happen after the namespace label enforcement, to avoid a tenant injecting metrics into another tenant.

To be clear: you're already setting `.spec.enforcedNamespaceLabel: "namespace"`? 
If yes, how do you deal with recording/alerting rules? Do you expect to enforce the tenant label there too?

simonpasquier commented:
> I forgot about those, I also need the tenant label for recording/alerting rules. 

:thinking: I don't see an obvious solution for rules since relabeling can't be applied here...

> Maybe a new `.spec.enforceMetricRelabelings` in the prometheus CRD would be more appropriate (please note that I don't want `excludedFromEnforcement` to be applied here).

I suppose that the latter is because you deploy generic rules generating namespace-aware metrics/alerts and you still want the metrics/alerts to have the right tenant?

There might still be other use cases that need to skip "enforced" metric relabelings for some monitor/rule resources though.

simonpasquier commented:
Nevertheless this PR (with scrape class' relabeling happening before monitor's metric relabeling) is worth the effort even if it's not addressing @sathieu requirements.

sathieu commented:
I've updated the PR to :

```go
	// The Operator adds the scrape class metric relabelings defined here.
	// Then the Operator adds the target-specific metric relabelings defined in the scrape object.
	// Then the Operator adds namespace enforcement relabeling rule.
```

I'll handle my usecase in another PR.

ArthurSens reviewed (commented):
Hi @sathieu, sorry for taking so long to give you an answer here, I saw the failing tests and assumed the PR was not ready. I took a deeper look today and I _think_ the failing tests are flaky ones. I've re-triggered them to be sure... let's see

sathieu commented:
Thanks @ArthurSens I've added your suggestion. I also had to rebase to resolve conflicts. Back to you!

nicolastakashi commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -1840,4 +1840,15 @@ type ScrapeClass struct {
 	// +optional
 	Relabelings []RelabelConfig `json:"relabelings,omitempty"`
+
+	// MetricRelabelings configures the relabeling rules to apply to all samples before injection.
```
Do you mean ingestion?

sathieu replied:
Yes :syringe: . Fixed, thanks.

ArthurSens commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -1840,4 +1840,15 @@ type ScrapeClass struct {
+	// MetricRelabelings configures the relabeling rules to apply to all samples before ingestion.
+	//
+	// The Operator adds the scrape class metric relabelings defined here.
+	// Then the Operator adds the target-specific metric relabelings defined in the scrape object.
```
```suggestion
	// Then the Operator adds the target-specific metric relabelings defined in ServiceMonitors, PodMonitors, Probes and ScrapeConfigs.
```

ArthurSens commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -1840,4 +1840,15 @@ type ScrapeClass struct {
+	//
+	// The Operator adds the scrape class metric relabelings defined here.
+	// Then the Operator adds the target-specific metric relabelings defined in the scrape object.
+	// Then the Operator adds namespace enforcement relabeling rule.
```
```suggestion
	// Then the Operator adds namespace enforcement relabeling rule, specified in '.spec.enforcedNamespaceLabel'.
```

ArthurSens commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -8783,3 +8783,205 @@ func TestPodMonitorWithNonDefaultScrapeClassRelabelings(t *testing.T) {
 	golden.Assert(t, string(cfg), "podMonitorObjectWithNonDefaultScrapeClassWithRelabelings.golden")
 }
+
+func TestServiceMonitorWithDefaultScrapeClassMetricRelabelings(t *testing.T) {
```
Awesome tests!

We could improve code readability here if we reduce code duplication using [table tests](https://dev.to/boncheff/table-driven-unit-tests-in-go-407b)

ArthurSens reviewed (dismissed):
Awesome, thanks for tackling this!

I've made a few nitpick comments, but if you want to focus on something I'd say improving test readability is what could receive some love right now.

Approving the PR because this seems to work as expected

ArthurSens commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -8783,3 +8783,205 @@ func TestPodMonitorWithNonDefaultScrapeClassRelabelings(t *testing.T) {
+		},
+	}
+
+	prometheus.Spec.ScrapeClasses = scrapeClasses
+	cg := mustNewConfigGenerator(t, prometheus)
+
+	cfg, err := cg.GenerateServerConfiguration(
+		context.Background(),
+		prometheus.Spec.EvaluationInterval,
+		prometheus.Spec.QueryLogFile,
+		prometheus.Spec.RuleSelector,
+		prometheus.Spec.Exemplars,
+		prometheus.Spec.TSDB,
+		prometheus.Spec.Alerting,
+		prometheus.Spec.RemoteRead,
+		map[string]*monitoringv1.ServiceMonitor{"monitor": serviceMonitor},
+		nil,
+		nil,
+		nil,
+		&assets.StoreBuilder{},
+		nil,
+		nil,
+		nil,
+		nil,
+	)
+	require.NoError(t, err)
+	golden.Assert(t, string(cfg), "serviceMonitorObjectWithDefaultScrapeClassWithMetricRelabelings.golden")
```
What I mean with code duplication is that this part is basically identical in all tests. We could have a single function with different inputs instead

```go
testCases := []struct{
  scrapeClass monitoringv1.ScrapeClass
  podMonitors map[string]monitoringv1.Podmonitor // can be nil
  serviceMonitors map[string]monitoringv1.ServiceMonitor // can be nil
  probes map[string]monitoringv1.Probe // can be nil
  scrapeConfigs map[string]monitoringv1alpha1.ScrapeConfig // can be nil
  goldenFile string  
}
```
You can put the implementation of the test inside a loop and asser that your inputs results are equal to the appropriate golden file

sathieu commented:
@nicolastakashi I've addressed your comments, rebased (there was a conflict because of 642d9cd3fe343251edd3219eeb9cd84c8a8a6856) and force-pushed. Back to you :basketball: !

ArthurSens reviewed (approved):
Perfect, thanks @sathieu !
